crates/bevy_render/src/view/window/screenshot.rs

```diff
@@ -0,0 +1,277 @@
+use std::{borrow::Cow, num::NonZeroU32, path::Path};
+
+use bevy_ecs::prelude::*;
+use bevy_log::info_span;
+use bevy_tasks::AsyncComputeTaskPool;
+use bevy_utils::HashMap;
+use bevy_window::WindowId;
+use parking_lot::Mutex;
+use thiserror::Error;
+use wgpu::{
+    CommandEncoder, Extent3d, ImageDataLayout, TextureFormat, COPY_BYTES_PER_ROW_ALIGNMENT,
+};
+
+use crate::{
+    prelude::Image,
+    render_resource::{BindGroup, BindGroupLayout, Buffer, RenderPipeline, Texture},
+    renderer::RenderDevice,
+    texture::TextureFormatPixelInfo,
+};
+
+use super::ExtractedWindows;
+
+pub type ScreenshotFn = Box<dyn FnOnce(Image) + Send + Sync>;
+
+/// A resource which allows for taking screenshots of the window.
+#[derive(Resource, Default)]
+pub struct ScreenshotManager {
+    // this is in a mutex to enable extraction with only an immutable reference
+    pub(crate) callbacks: Mutex<HashMap<WindowId, ScreenshotFn>>,
+}
+
+#[derive(Error, Debug)]
+#[error("A screenshot for this window has already been requested.")]
+pub struct ScreenshotAlreadyRequestedError;
+
+impl ScreenshotManager {
+    /// Signals the renderer to take a screenshot of this frame.
+    ///
+    /// The given callback will eventually be called on one of the [`AsyncComputeTaskPool`]s threads.
+    pub fn take_screenshot(
+        &mut self,
+        window: WindowId,
+        callback: impl FnOnce(Image) + Send + Sync + 'static,
+    ) -> Result<(), ScreenshotAlreadyRequestedError> {
+        self.callbacks
+            .get_mut()
+            .try_insert(window, Box::new(callback))
+            .map(|_| ())
+            .map_err(|_| ScreenshotAlreadyRequestedError)
+    }
+
+    /// Signals the renderer to take a screenshot of this frame.
+    ///
+    /// The screenshot will eventually be saved to the given path, and the format will be derived from the extension.
+    pub fn save_screenshot_to_disk(
+        &mut self,
+        window: WindowId,
+        path: impl AsRef<Path>,
+    ) -> Result<(), ScreenshotAlreadyRequestedError> {
+        let path = path.as_ref().to_owned();
+        self.take_screenshot(window, |image| {
+            image.try_into_dynamic().unwrap().save(path).unwrap();
```

Comment:
would it be possible to bubbleup the error or at least use expect here? I feel like hitting a generic image conversion error might be confusing if someone is just trying to take a screenshot and it fails here.